agent/internal/core.go

```diff
@@ -0,0 +1,83 @@
+package internal
+
+import (
+	"context"
+	"fmt"
+	"os/exec"
+	"os/signal"
+	"syscall"
+	"time"
+
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+
+	"github.com/determined-ai/determined/agent/internal/options"
+	"github.com/determined-ai/determined/master/pkg/groupx/errgroupx"
+)
+
+// Run runs a new agent system and actor with the provided options.
+func Run(parent context.Context, version string, opts options.AgentOptions) error {
+	ctx, stop := signal.NotifyContext(parent, syscall.SIGINT, syscall.SIGTERM)
+	defer stop()
+
+	printableConfig, err := opts.Printable()
+	if err != nil {
+		return err
+	}
+	log.Infof("agent configuration: %s", printableConfig)
+
+	wg := errgroupx.WithContext(ctx)
+
+	log.Trace("starting main agent process")
+	wg.Go(func(ctx context.Context) error {
+		connectionFailureWindowBegin := time.Now()
+		connectionFailureCount := 0
+		for {
+			a := New(ctx, version, opts)
```

Comment:
yep. i added this back because the agent actually needs to distinguish connection lost from connection blipped - it was slightly cleaner before but this is ok.